File: clients/ios/CuraVeris/SimCardModule.m
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include("RCTBridgeModule.h")
#import "RCTBridgeModule.h"
#endif

#ifdef RCT_EXTERN_MODULE
RCT_EXTERN_MODULE(SimCardModule, NSObject)

RCT_EXTERN_METHOD(
  getSimCards:(RCTPromiseResolveBlock)resolve
  rejecter:(RCTPromiseRejectBlock)reject
)

@end
#endif

File: clients/ios/CuraVeris/SimCardModule.swift
import Foundation
import CoreTelephony

@objc(SimCardModule)
class SimCardModule: NSObject {

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func getSimCards(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let networkInfo = CTTelephonyNetworkInfo()

    var simCards: [[String: Any]] = []

    if let providers = networkInfo.serviceSubscriberCellularProviders {
      for (key, carrier) in providers {
        var simInfo: [String: Any] = [:]
        simInfo["slotKey"] = key
        simInfo["carrierName"] = carrier.carrierName ?? ""
        simInfo["mobileCountryCode"] = carrier.mobileCountryCode ?? ""
        simInfo["mobileNetworkCode"] = carrier.mobileNetworkCode ?? ""
        simInfo["isoCountryCode"] = carrier.isoCountryCode ?? ""
        simInfo["allowsVOIP"] = carrier.allowsVOIP
        simInfo["number"] = ""
        simCards.append(simInfo)
      }
    }

    resolve(simCards)
  }
}
